onnxruntime/core/session/inference_session.cc

```diff
@@ -415,27 +415,39 @@ class InferenceSession::Impl {
                   "Unexpected input data type. Actual: (" + actual_name + ") , expected: (" + expected_name + ")");
   }
 
-  common::Status ValidateInputTypes(const std::vector<std::string>& feed_names,
-                                    const std::vector<MLValue>& feeds) {
+  common::Status ValidateInputs(const std::vector<std::string>& feed_names,
+                                const std::vector<MLValue>& feeds) {
+    if (required_input_def_list_.size() != feed_names.size() || feed_names.size() != feeds.size()) {
+      std::ostringstream ostr;
+      std::for_each(std::begin(required_model_input_names_),
+                    std::end(required_model_input_names_),
+                    [&ostr](const std::string& elem) {
+                      ostr << elem << " ";
+                    });
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                             "Invalid inputs. Valid input names are: ", ostr.str());
```

Comment:
>Valid [](start = 46, length = 5)

nit: I think the error message should be a bit more verbose. If I read the ONNX spec a 'valid' input to me would be anything in the GraphProto input list (which includes the initializer names). We should say that the list is the inputs we require, and that no others are accepted because we don't allow overriding initializers.